### PR TITLE
Threading rework + bonuses

### PR DESCRIPTION
**Pull request by lafreak**

### CODE REVIEW

File: apps/gameserver/include/World.h [hunk 1]
```diff
@@ -235,12 +293,17 @@ class World
...
+        std::lock_guard<std::recursive_mutex> lock(Get().m_entities_rmtx);
         {
-            std::lock_guard<std::recursive_mutex> lock(Get().m_entities_rmtx);
-            Get().m_players.insert(std::make_pair(entity->GetID(), entity));
+            // FIXME: Don't we need to lock for the entire time of adding to the map as well?
+            //std::lock_guard<std::recursive_mutex> lock(Get().m_entities_rmtx);
+            auto [_, inserted] = Get().m_players.insert(std::make_pair(entity->GetID(), entity));
+            if (!inserted)
+                throw std::logic_error("player already present in world"); 
         }
```
Comment (Cr3ativeZen) [inline apps/gameserver/include/World.h:307]: This brackets are not needed anymore.

File: apps/gameserver/include/World.h [hunk 2]
```diff
@@ -251,21 +314,34 @@ class World
...
-                Get().m_loots.insert(std::make_pair(entity->GetID(), std::dynamic_pointer_cast<Loot>(entity)));
+            {
+                auto [_, inserted] = Get().m_loots.insert(std::make_pair(entity->GetID(), std::dynamic_pointer_cast<Loot>(entity)));
+                if (!inserted)
+                    throw std::logic_error("loot already present in world"); 
                 break;
             }
+            }
 
         }
```
Comment (Cr3ativeZen) [inline apps/gameserver/include/World.h:346]: These brackets are not needed aswell.

File: apps/gameserver/include/World.h [hunk 3]
```diff
@@ -280,24 +356,37 @@ class World
...
     //! Removes player from the world.
     static void Remove(Player* entity)
     {
+        std::lock_guard<std::recursive_mutex> lock(Get().m_entities_rmtx);
         Map(entity->GetMap()).Remove(entity);
         {
-            std::lock_guard<std::recursive_mutex> lock(Get().m_entities_rmtx);
+            // std::lock_guard<std::recursive_mutex> lock(Get().m_entities_rmtx);
             Get().m_players.erase(entity->GetID());
         }
```
Comment (Cr3ativeZen) [inline apps/gameserver/include/World.h:367]: brackets and comment inside it.

File: apps/gameserver/include/World.h [hunk 4]
```diff
@@ -280,24 +356,37 @@ class World
...
     }
 
+    static void Remove(Loot* entity)
+    {
+        std::lock_guard<std::recursive_mutex> lock(Get().m_entities_rmtx);
+        Map(entity->GetMap()).Remove(entity);
+        {
+            // std::lock_guard<std::recursive_mutex> lock(Get().m_entities_rmtx);
+            Get().m_loots.erase(entity->GetID());
+        }
```
Comment (Cr3ativeZen) [inline apps/gameserver/include/World.h:377]: brackets and comment

File: apps/gameserver/src/BeheadableMonster.cpp [hunk 5]
```diff
@@ -17,7 +17,7 @@ void BeheadableMonster::Die()
 
 void BeheadableMonster::Tick()
 {
-    if(IsGState(CGS_KNEE) && (time::now() - GetDeathTime()).count() >= 10000)
+    if(IsGState(CGS_KNEE) && (time::now() - GetDeathTime()).count() >= 1000)//10000) // FIXME: Change back when behead is added.
```
Comment (Cr3ativeZen) [inline apps/gameserver/src/BeheadableMonster.cpp:20]: We should use some static variable to hold the time.

File: apps/gameserver/src/Character.cpp [hunk 6]
```diff
@@ -130,7 +140,8 @@ void Character::ReceiveDamage(id_t id, std::uint32_t damage)
 
 void Character::WriteInSight(const packet& p) const
 {
-    World::Map(GetMap()).WriteInSight(*this, p);
+    // World::Map(GetMap()).WriteInSight(*this, p);
```
Comment (Cr3ativeZen) [inline apps/gameserver/src/Character.cpp:143]: is this comment necessary?

File: apps/gameserver/src/Monster.cpp [hunk 7]
```diff
@@ -168,7 +169,7 @@ void Monster::ReceiveDamage(id_t id, std::uint32_t damage)
 void Monster::DistributeExp()
 {
     std::map<id_t, std::uint64_t> party_container;
-    for (auto&[id, damage] : hostility_map)
+    for (auto& [id, damage] : hostility_map)
```
Comment (Cr3ativeZen) [inline apps/gameserver/src/Monster.cpp:172]: I thnik we can set it as const.

File: apps/gameserver/src/Player.cpp [hunk 8]
```diff
@@ -633,7 +643,8 @@ void Player::OnAttack(packet& p)
     auto id = p.pop<Character::id_t>();
     auto z = p.pop<unsigned int>();
 
-    World::Map(GetMap()).For(WorldMap::QK_PLAYER | WorldMap::QK_MONSTER, id, [&](Character& character) 
+    //World::Map(GetMap()).For(WorldMap::QK_PLAYER | WorldMap::QK_MONSTER, id, [&](Character& character)
```
Comment (Cr3ativeZen) [inline apps/gameserver/src/Player.cpp:646]: Is this comment needed?

File: apps/gameserver/src/Player.cpp [hunk 9]
```diff
@@ -662,9 +673,15 @@ void Player::OnAttack(packet& p)
         LookAt(&character);
 
         auto defender_lock = character.Lock();
+        if (!CanAttack(character))
+        {
+            spdlog::debug("Player {} cannot attack monster character id {} due to some reason",
+                this->GetName(), character.GetID());
```
Comment (Cr3ativeZen) [inline apps/gameserver/src/Player.cpp:679]: Would be nice to see what the actual reason is. This log could be moved to CanAttack.

File: apps/gameserver/src/Player.cpp [hunk 10]
```diff
@@ -787,7 +804,9 @@ void Player::OnItemPick(packet& p)
     World::ForLoot(item_id, [&](Loot& loot) {
                     auto& info = loot.GetItemInfo();
                     InsertItem(info.Index, info.Num);
-                    World::RemoveLootById(item_id);
+                    spdlog::debug("Player {} picked item index {} num {}", GetName(), info.Index, info.Num);
+                    //World::RemoveLootById(item_id);
```
Comment (Cr3ativeZen) [inline apps/gameserver/src/Player.cpp:808]: comment

File: apps/gameserver/src/Player.cpp [hunk 11]
```diff
@@ -844,6 +863,14 @@ void Player::BanFromParty(id_t banned_player_id)
 
 void Player::Tick()
 {
+    auto now = time::now();
+    if ((now - m_last_save).count() > 60'000)
```
Comment (Cr3ativeZen) [inline apps/gameserver/src/Player.cpp:867]: We could use some static variable here.

File: apps/gameserver/src/main.cpp [hunk 12]
```diff
@@ -193,7 +194,8 @@ int main(int argc, char** argv)
         spdlog::info("/around radius: {}; player {}:", radius, player.GetName());
 
         auto query = WorldMap::QK_PLAYER|WorldMap::QK_MONSTER|WorldMap::QK_NPC;
-        World::Map(player.GetMap()).ForEachAround(player, radius, query, [&](Character& character) {
+        //World::Map(player.GetMap()).ForEachAround(player, radius, query, [&](Character& character) {
```
Comment (Cr3ativeZen) [inline apps/gameserver/src/main.cpp:197]: comment

File: apps/gameserver/include/World.h [hunk 13]
```diff
@@ -251,21 +314,34 @@ class World
         if (entity->GetType() == Character::PLAYER)
             throw std::logic_error("player should be added to the world by reference");
 
+        std::lock_guard<std::recursive_mutex> lock(Get().m_entities_rmtx);
         {
-            std::lock_guard<std::recursive_mutex> lock(Get().m_entities_rmtx);
+            //std::lock_guard<std::recursive_mutex> lock(Get().m_entities_rmtx);
```
Comment (Cr3ativeZen) [inline apps/gameserver/include/World.h:319]: comment

File: apps/gameserver/src/Player.cpp (hunk 9 above)
Comment (lafreak) [reply apps/gameserver/src/Player.cpp:679]: In future there will be bunch of reasons, Do we really need to log each case?

File: apps/gameserver/include/World.h [hunk 14]
```diff
@@ -13,7 +13,7 @@
 #include <bango/utils/time.h>
 
 #define MAP_WIDTH 50*8192
-#define MAP_SIGHT 1024
+#define MAP_SIGHT 920//1024
```
[the remark this answers is not in the document]
Comment (lafreak) [reply apps/gameserver/include/World.h:16]: I played around with this value to get rid of issue with bots reporting appear packets being sent from too far away. I'm not sure whats the original value inside inix binary but lowering to 920 worked for me.

Code inside kal-ai:
```python
        # This is workaround for common issues with incoming old place appear packets after teleportation
        if x_diff > 1000 or y_diff > 1000:
            self.logger.warning(f"Player appear ignored: xy_diff:({x_diff:5},{y_diff:5})")
            return
```

File: apps/gameserver/include/Player.h [hunk 15]
```diff
@@ -23,14 +23,14 @@ class Player : public Character, public User
     Inventory m_inventory;
     std::shared_ptr<Party> m_party;
     int m_party_inviter_id;
-    id_t m_party_id;
+    id_t m_party_id=0;
```
[the remark this answers is not in the document]
Comment (lafreak) [reply apps/gameserver/include/Player.h:26]: Fixed obvious bug with `m_party_inviter_id`, did not touch other classes. Let's initialize uninitialized ones in separate task.

File: apps/gameserver/src/Player.cpp (hunk 9 above)
Comment (Cr3ativeZen) [reply apps/gameserver/src/Player.cpp]: true.

File: apps/gameserver/include/Player.h (hunk 15 above)
[the remark this answers is not in the document]
Comment (Cr3ativeZen) [reply apps/gameserver/include/Player.h:26]: Please create this task.

**Review by Cr3ativeZen — APPROVED** [review]

**Pull request merged**